Errors when trying to use Docker to build

#### Version of the game
commit id: c966395429a204701ac54e17710a4f4eab8fc709

#### What is the bug?
I saw ~4 different build issues when I followed the Docker build instructions in a brand new default "blank" Ubuntu 23.04 VM. I don't know if I should've opened a new issue for each one or just added comments to #409 which is pretty similar (this one is possibly at least partially a duplicate?) but decided that this might be different enough and there's enough different stuff and details to just open a new issue. Feel free to close if this is not helpful.

It would be great if someone could replicate my steps below as exactly as possible and report (in great detail) if they get different results.

#### Steps to reproduce the behavior:

"Step 0": here's my setup, ideally you would have something that's as close as possible:
Host: Debian 12.4 (x86_64 Linux computer)
I used VirtualBox 7.0.12 to make and run the VM
New blank Ubuntu 23.04 VM (I used [ubuntu-23.04-desktop-legacy-amd64.iso](https://cdimage.ubuntu.com/releases/lunar/release/ubuntu-23.04-desktop-legacy-amd64.iso))
I gave it 4 GB RAM, 2 cores, 60 GB hard disk, and when installing Ubuntu, I chose default everything, except "Minimal Installation" in Ubuntu installer

In the VM, first ever login, after rejecting everything Ubuntu/GNOME offered by default:

(Completely new VM, don't do absolutely anything else than the following exact steps)
1. `sudo apt update && sudo apt upgrade && sudo apt autoremove`
2. `sudo apt install git make`
3. `git clone https://github.com/lambertjamesd/portal64`
(commit c966395429a204701ac54e17710a4f4eab8fc709)
4. `cd portal64`
5. (next copy the required Portal files to vpk/)
6. `sudo apt install docker.io`
7. `sudo setfacl -R -m u:$USER:rwx /var/run/docker.sock`
I jumped straight to the Docker portion of the README, skipped everything before that
8. `make -f Makefile.docker image`
9. `make -f Makefile.docker`

After that, I got my first build error:

```
make[1]: Entering directory '/usr/src/app/skelatool64'
g++ -Wall -Werror -g -rdynamic -I./yaml-cpp/include  -c src/lua_generator/LuaYaml.cpp -o build/src/lua_generator/LuaYaml.o
src/lua_generator/LuaYaml.cpp:3:10: fatal error: yaml-cpp/yaml.h: No such file or directory
    3 | #include "yaml-cpp/yaml.h"
      |          ^~~~~~~~~~~~~~~~~
compilation terminated.
```

I think this was already a bug: the README can kind of be kind of be interpreted that you don't have to perform any of the other steps before following the Docker section, so I blindly skipped directly to the `make -f Makerfile.docker` stuff (except I knew that I must first copy the Portal vpk/ files)

I tried to fix the situation like this:
10. `cd skelatool64`
11. `./setup_dependencies.sh`

that said among other things:

```
-- Build files have been written to: /home/user/portal64/skelatool64/yaml-cpp
```

(bug: `make -f Makefile.docker` should have automatically run `skelatool64/setup_dependencies.sh`, right?)

Now that the necessary files were generated (by having to manually run `./setup_dependencies.sh` in Step 11), I tried to continue normally:

(I didn't `make` "outside of Docker")
12. `cd ..`
13. `make -f Makefile.docker`

No errors reported this time, but the ROM was buggy (see below for more details).

#### The bugs:

So I guess these are what I consider bugs in what happened above:

1) the issue after step 9: "`yaml-cpp/yaml.h: No such file or directory"` etc because `make -f Makefile.docker` should have automatically run `skelatool64/setup_dependencies.sh` (?)

2) The build was "successful" but the resulting ROM didn't work. There should have been an error message somewhere because there's something wrong with the ROM.

3) The ROM looked superficially fine, but it compressed to ~900 KB which is obviously too small (something similar was also reported in the comments of #409, possibly audio related or something), and the screen was completely blank when I tried to run it in an emulator.

4) Also, I found it surprising that what Docker built was owned by root, not the user:
- build/
- portal_pak_dir/
- portal_pak_modified/
- skelatool64/build/
- skelatool64/skeletool

Can the current username that ran the docker command be made the owner of those directories and files instead? It's weird that root owns files in my home directory.

You dont need to do anything with skelatool if you use docker, everything
gets build inside the docker container.

The main Makefile builds skelatool already automatically and also executes
setup dependencies.

If the rom compresses that much, it does not include needed data. I dont
know how that can happen.

Whats your file system structure in vpk/ ?

I think its normal that the files built by docker are owned by root, thats
why the original version of the docker makefile used sudo. I guess the
docker service rans as root?

But I dont know Docker, so I dont know how it is supposed to be...?

I mean generally it works fine, my fresh Docker build works, so until I can
reproduce a problematic enviroment as yours, maybe you could debug why it
does not work? :)


rxhfcy ***@***.***> schrieb am Mi., 13. Dez. 2023, 00:05:

> Version of the game
>
> commit id: c966395
> <https://github.com/lambertjamesd/portal64/commit/c966395429a204701ac54e17710a4f4eab8fc709>
> What is the bug?
>
> I saw ~4 different build issues when I followed the Docker build
> instructions in a brand new default "blank" Ubuntu 23.04 VM. I don't know
> if I should've opened a new issue for each one or just added comments to
> #409 <https://github.com/lambertjamesd/portal64/issues/409> which is
> pretty similar (this one is possibly at least partially a duplicate?) but
> decided that this might be different enough and there's enough different
> stuff and details to just open a new issue. Feel free to close if this is
> not helpful.
>
> It would be great if someone could replicate my steps below as exactly as
> possible and see if they get different results.
> Steps to reproduce the behavior:
>
> "Step 0": here's my setup, ideally you would have something that's as
> close as possible:
> Host: Debian 12.4 (x86_64 Linux computer)
> I used VirtualBox 7.0.12 to make and run the VM
> New blank Ubuntu 23.04 VM (I used ubuntu-23.04-desktop-legacy-amd64.iso
> <https://cdimage.ubuntu.com/releases/lunar/release/ubuntu-23.04-desktop-legacy-amd64.iso>
> )
> I gave it 4 GB RAM, 2 cores, 60 GB hard disk, and when installing Ubuntu,
> I chose default everything, except "Minimal Installation" in Ubuntu
> installer
>
> In the VM, first ever login, after rejecting everything Ubuntu/GNOME
> offered by default:
>
> (Completely new VM, don't do absolutely anything else than the following
> exact steps)
>
>    1. sudo apt update && sudo apt upgrade && sudo apt autoremove
>    2. sudo apt install git make
>    3. git clone https://github.com/lambertjamesd/portal64
>    (commit c966395
>    <https://github.com/lambertjamesd/portal64/commit/c966395429a204701ac54e17710a4f4eab8fc709>
>    )
>    4. cd portal64
>    5. (next copy the required Portal files to vpk/)
>    6. sudo apt install docker.io
>    7. sudo setfacl -R -m u:$USER:rwx /var/run/docker.sock
>    I jumped straight to the Docker portion of the README, skipped
>    everything before that
>    8. make -f Makefile.docker image
>    9. make -f Makefile.docker
>
> After that, I got my first build error:
>
> make[1]: Entering directory '/usr/src/app/skelatool64'
> g++ -Wall -Werror -g -rdynamic -I./yaml-cpp/include  -c src/lua_generator/LuaYaml.cpp -o build/src/lua_generator/LuaYaml.o
> src/lua_generator/LuaYaml.cpp:3:10: fatal error: yaml-cpp/yaml.h: No such file or directory
>     3 | #include "yaml-cpp/yaml.h"
>       |          ^~~~~~~~~~~~~~~~~
> compilation terminated.
>
> I think this was already a bug: the README can kind of be kind of be
> interpreted that you don't have to perform any of the other steps before
> following the Docker section, so I blindly skipped directly to the make
> -f Makerfile.docker stuff (except I knew that I must first copy the
> Portal vpk/ files)
>
> I tried to fix the situation like this:
> 10. cd skelatool64
> 11. ./setup_dependencies.sh
>
> that said among other things:
> -- Build files have been written to:
> /home/user/portal64/skelatool64/yaml-cpp
> (bug: make -f Makefile.docker should have automatically run
> skelatool64/setup_dependencies.sh, right?)
>
> Now that the build files had been generated, I tried to continue normally:
>
> (I didn't make "outside of Docker")
> 12. cd ..
> 13. make -f Makefile.docker
>
> No errors reported this time, but the ROM was buggy (see below for more
> details).
> The bugs:
>
> So I guess these are what I consider bugs in what happened above:
>
>    1.
>
>    the issue after step 9: "yaml-cpp/yaml.h: No such file or directory"
>    etc because make -f Makefile.docker should have automatically run
>    skelatool64/setup_dependencies.sh (?)
>    2.
>
>    The build was "successful" but the resulting ROM didn't work. There
>    should have been an error message somewhere because there's something wrong
>    with the ROM.
>    3.
>
>    The ROM looked superficially fine, but it compressed to ~900 KB which
>    is obviously too small (something similar was also reported in the comments
>    of #409 <https://github.com/lambertjamesd/portal64/issues/409>,
>    possibly audio related or something), and the screen was completely blank
>    when I tried to run it in an emulator.
>    4.
>
>    Also, I found it surprising that what Docker built was owned by root,
>    not the user:
>
>
>    - build/
>    - portal_pak_dir/
>    - portal_pak_modified/
>    - skelatool64/build/
>    - skelatool64/skeletool
>
> Can the current username that ran the docker command be made the owner of
> those directories and files instead? It's weird that root owns files in my
> home directory.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/lambertjamesd/portal64/issues/555>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAANGLDIJQGACQRAQSNSGRTYJDPK5AVCNFSM6AAAAABASHRDFOVHI2DSMVQWIX3LMV43ASLTON2WKOZSGAZTQNRXGY4TANI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

--- a/Makefile
+++ b/Makefile
@@ -69,11 +69,12 @@ LDFLAGS =	-L/usr/lib/n64 $(N64LIB)  -L$(N64_LIBGCCDIR) -lgcc
 
 default:	english_audio
 
-english_audio: build/src/audio/subtitles.h portal_pak_dir
+english_audio: build/src/audio/subtitles.h portal_pak_dir $(SKELATOOL64)
 	@$(MAKE) -C skelatool64
 	@$(MAKE) buildgame
 
-all_languages: build/src/audio/subtitles.h portal_pak_dir german_audio french_audio russian_audio spanish_audio
+all_languages: build/src/audio/subtitles.h portal_pak_dir german_audio french_audio russian_audio spanish_audio $(SKELATOOL64)
+	@$(MAKE) -C skelatool64
 	@$(MAKE) buildgame
 
 german_audio: vpk/portal_sound_vo_german_dir.vpk vpk/portal_sound_vo_german_000.vpk portal_pak_dir

--- a/skelatool64/makefile
+++ b/skelatool64/makefile
@@ -34,7 +34,10 @@ skeletool64: $(OBJ_FILES) $(LUA_OBJ_FILES)
 clean:
 	rm -rf build/
 	rm -f skeletool64
-	rm -f CImg_*
+	rm -f *.zip*
+	rm -rf CImg-*
+	rm -rf cimg
+	@$(MAKE) -C yaml-cpp clean
 
 init:
 	

--- a/skelatool64/setup_dependencies.sh
+++ b/skelatool64/setup_dependencies.sh
@@ -9,8 +9,9 @@ git clone https://github.com/jbeder/yaml-cpp.git
 cmake -S yaml-cpp -B yaml-cpp
 make -C yaml-cpp
 
-wget http://cimg.eu/files/CImg_3.1.3.zip
-unzip CImg_3.1.3.zip
+wget -O CImg_3.1.3.zip http://cimg.eu/files/CImg_3.1.3.zip
+unzip -o CImg_3.1.3.zip
+rm -rf cimg
 mv CImg-3.1.3 cimg
 
 popd

